src/jitlayers.cpp

```diff
@@ -0,0 +1,396 @@
+    typedef CompileLayerT::ModuleSetHandleT ModuleHandleT;
+    typedef StringMap<void*> GlobalSymbolTableT;
+    typedef object::OwningBinary<object::ObjectFile> OwningObj;
+
+    JuliaOJIT(TargetMachine &TM)
+      : TM(TM),
+        DL(TM.createDataLayout()),
+        ObjStream(ObjBufferSV),
+        MemMgr(
+#ifdef CUSTOM_MEMORY_MANAGER
+            createRTDyldMemoryManagerOSX()
+#else
+            new SectionMemoryManager
+#endif
+            ) {
+#ifdef JL_DEBUG_BUILD
+            PM.add(createVerifierPass());
+#endif
+            // In imaging mode, we run the pass manager on creation
+            // to make sure it ends up optimized in the shadow module
+            if (!imaging_mode) {
+                addOptimizationPasses(&PM);
+#ifdef JL_DEBUG_BUILD
+                PM.add(createVerifierPass());
+#endif
+            }
+            if (TM.addPassesToEmitMC(PM, Ctx, ObjStream))
+                llvm_unreachable("Target does not support MC emission.");
+
+            CompileLayer = std::unique_ptr<CompileLayerT>{new CompileLayerT(ObjectLayer,
+                [&](Module &M) {
+                    PM.run(M);
+                    std::unique_ptr<MemoryBuffer> ObjBuffer(
+                        new ObjectMemoryBuffer(std::move(ObjBufferSV)));
+                    ErrorOr<std::unique_ptr<object::ObjectFile>> Obj =
+                        object::ObjectFile::createObjectFile(ObjBuffer->getMemBufferRef());
+
+                    // TODO: Actually report errors helpfully.
+                    if (Obj)
+                        return OwningObj(std::move(*Obj), std::move(ObjBuffer));
+                    return OwningObj(nullptr, nullptr);
+                }
+            )};
+            // Make sure SectionMemoryManager::getSymbolAddressInProcess can resolve
+            // symbols in the program as well. The nullptr argument to the function
+            // tells DynamicLibrary to load the program, not a library.
+
+            std::string *ErrorStr = nullptr;
+            if (sys::DynamicLibrary::LoadLibraryPermanently(nullptr, ErrorStr))
+                report_fatal_error("FATAL: unable to dlopen self\n" + *ErrorStr);
+        }
+
+    std::string mangle(const std::string &Name) {
+        std::string MangledName;
+        {
+            raw_string_ostream MangledNameStream(MangledName);
+            Mangler::getNameWithPrefix(MangledNameStream, Name, DL);
+        }
+        return MangledName;
+    }
+
+    void addGlobalMapping(StringRef Name, void *Addr) {
+       GlobalSymbolTable[mangle(Name)] = Addr;
+    }
+
+    ModuleHandleT addModule(Module *M) {
+        // We need a memory manager to allocate memory and resolve symbols for this
+        // new module. Create one that resolves symbols by looking back into the
+        // JIT.
+        auto Resolver = orc::createLambdaResolver(
+                          [&](const std::string &Name) {
+                            // TODO: consider moving the FunctionMover resolver here
```

Comment:
I think it might still be reasonable to change the way we do bootstrapping once we get rid of MCJIT